src/api/core/accounts.rs

```diff
@@ -770,15 +798,22 @@ async fn prelogin(data: JsonUpcase<PreloginData>, conn: DbConn) -> Json<Value> {
 pub async fn _prelogin(data: JsonUpcase<PreloginData>, mut conn: DbConn) -> Json<Value> {
     let data: PreloginData = data.into_inner().data;
 
-    let (kdf_type, kdf_iter) = match User::find_by_mail(&data.Email, &mut conn).await {
-        Some(user) => (user.client_kdf_type, user.client_kdf_iter),
-        None => (User::CLIENT_KDF_TYPE_DEFAULT, User::CLIENT_KDF_ITER_DEFAULT),
+    let (kdf_type, kdf_iter, kdf_mem, kdf_para) = match User::find_by_mail(&data.Email, &mut conn).await {
+        Some(user) => (user.client_kdf_type, user.client_kdf_iter, user.client_kdf_memory, user.client_kdf_parallelism),
+        None => (User::CLIENT_KDF_TYPE_DEFAULT, User::CLIENT_KDF_ITER_DEFAULT, None, None),
     };
 
-    Json(json!({
+    let mut result = json!({
         "Kdf": kdf_type,
-        "KdfIterations": kdf_iter
-    }))
+        "KdfIterations": kdf_iter,
+    });
+
+    if kdf_type == UserKdfType::Argon2id as i32 {
+        result["KdfMemory"] = Value::Number(kdf_mem.unwrap().into());
+        result["KdfParallelism"] = Value::Number(kdf_para.unwrap().into());
+    }
```

Comment:
I think using if let Some(x) = Option<> syntax would be a lot better than unwrapping.
```suggestion
    if let Some(m) = kdf_mem {
        result["KdfMemory"] = Value::Number(m.into())
    }
    if let Some(p) = kdf_para {
        result["KdfParallelism"] = Value::Number(p.into())
    }
```

Reply:
Then, instead of an `unwrap()` i suggest to use a `expect()` instead, and provide a correct message in case it fails.
Now it's probably guessing.

Reply:
This will make the code look like crap and cluttered. See the comment from djinn in another thread.

If there are 3 people who are reviewing the PR and they have different opinions, this PR will never be finished. I am not against suggestions and help, especially since I definitely need help with Rust patterns, but can you plese agree on an approach.

I have fixed all the outstanding issues and now it seems (at least to me) there is a contradictory suggestion than to what I got before.

Reply:
There are 4 options right now:

1. leave it as it is
2. replace `unwrap` with `expect`
3. use BlockListed's suggestion
4. change it to the following (IMO the best option):

```rust
    if kdf_type == UserKdfType::Argon2id as i32 {
        match kdf_mem {
            Some(m) => {
                result["KdfMemory"] = Value::Number(m.into());
            }
            None => panic!("KdfMemory must not be empty, when Argon2 is used."),
        }
        match kdf_para {
            Some(p) => {
                result["KdfParallelism"] = Value::Number(p.into());
            }
            None => panic!("KdfParallelism must not be empty, when Argon2 is used."),
        }
    }
```

Reply:
please let me know which one to pick and I make the changes. I favor option number 4.

Reply:
Panicking seems like the best solution and `unwrap`ping is not really intended for production code in rust, so the most rusty thing to choose would be `expect` in my opinion.

Reply:
If you are only going to use `match` to send an `panic` which i would actually not want to be used at all, then an `except` would be the better choice. So, i agree with @BlockListed

Reply:
ok, `expect` it is.